### PR TITLE
fix: add url encoding for multi-byte artifact name by using curl option

### PR DESCRIPTION
**Pull request by sakka2**
## Context
When we use multibyte character at the name of artifacts, it would be error because of non-encoded url. So I added url encoding before uploading artifacts to the store.

## Objective
- Add url encoding when uploading artifacts to the store.
- Use curl option "--data-urlencode" for encoding.
- "--data-urlencode" option puts "/?" at the beginning, so I eliminated.
- Use sd-step for curl only when curl version is old (before "7.18.0") and cannot use "--data-urlencode" option.

## References
Related to: https://github.com/screwdriver-cd/screwdriver/pull/942  
Reference: https://github.com/screwdriver-cd/scm-github/blob/bac74f62cb1bc937b4bb3371ef74f866b20d0313/index.js#L274
About curl option: https://curl.haxx.se/docs/manpage.html

## Test in builds
I have tested in my screwdriver environment.
Curl version of my environment is `7.29.0`, so the test below assumes that in case curl version is before `7.29.0`, `sd-step` will be used and in case after `7.29.0`, `sd-step` will not be used.
Both case, url encoding has succeeded.

### CODE REVIEW

File: commands.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
 cd $SD_ARTIFACTS_DIR
 find . -type f -print > manifest.txt
-find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; curl -H "Authorization: Bearer $SD_TOKEN" -H "Content-Type: text/plain" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$f -T ./$f' -- {} \;
+find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; encoded=`curl -s -w "%{url_effective}\n" --data-urlencode $f -G "" | sed -e "s/^\/?//"`; curl -H "Authorization: Bearer $SD_TOKEN" -H "Content-Type: text/plain" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$encoded -T ./$f' -- {} \;
```
Comment (tk3fftk) [inline commands.txt]: I think some image including old `curl` (before "7.18.0") cannot use `--data-urlencode` option.
https://curl.haxx.se/docs/manpage.html

Comment (sakka2) [reply commands.txt]: Oh, I see.
So it's better to use sd-step in order to use latest version.

File: commands.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
 cd $SD_ARTIFACTS_DIR
 find . -type f -print > manifest.txt
-find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; curl -H "Authorization: Bearer $SD_TOKEN" -H "Content-Type: text/plain" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$f -T ./$f' -- {} \;
+find . -type f -exec sh -c 'f=`echo $1 | sed "s/^\.\///"`; encoded=`sd-step exec core/curl "curl -s -w \"%{url_effective}\n\" --data-urlencode $f -G \"\" | sed -e \"s/^\/?//\""`; curl -H "Authorization: Bearer $SD_TOKEN" -H "Content-Type: text/plain" -X PUT $STORE_URL/v1/builds/$SD_BUILD_ID/$ARTIFACTS_DIR_SUFFIX/$encoded -T ./$f' -- {} \;
```
Comment (tk3fftk) [inline commands.txt]: I think using `sd-step` every time may take long time so this should be like scm-github https://github.com/screwdriver-cd/scm-github/blob/bac74f62cb1bc937b4bb3371ef74f866b20d0313/index.js#L274

Comment (sakka2) [reply commands.txt]: I fixed.
Only when `--data-urlencode` option cannot be used, exec sd-step command.

**Review by yuichi10 — APPROVED** [review]
It looks work well

**Review by tk3fftk — APPROVED** [review]
LGTM:+1:

**Comment by tkyi** [comment]
Have you tried running it in a build?

**Comment by sakka2** [comment]
Yes, I have tried in our screwdriver environment.
Please check the test result.

**Comment by tkyi** [comment]
Okay, LGTM then! 👍

**Review by tkyi — APPROVED** [review]

**Review by FenrirUnbound — APPROVED** [review]
👍 

rainy day somewhere in the caribbean - imgur

**Pull request merged**